floris/tools/wind_data.py

```diff
@@ -0,0 +1,529 @@
+            _,
+            freq_table_unpack,
+            _,
+            _,
+        ) = self.unpack()
+
+        return freq_table_unpack
+
+
+class WindRose(WindDataBase):
+    """
+    In FLORIS v4, the WindRose class is used to drive FLORIS and optimization
+    operations in which the inflow is characterized by the frequency of
+    binned wind speed, wind direction and turbulence intensity values
+
+    Args:
+        wind_directions: NumPy array of wind directions (NDArrayFloat).
+        wind_speeds: NumPy array of wind speeds (NDArrayFloat).
+        freq_table: Frequency table for binned wind direction, wind speed
+            values (NDArrayFloat, optional).  Defaults to None.
+        ti_table: Turbulence intensity table for binned wind direction, wind
+            speed values (NDArrayFloat, optional).  Defaults to None.
+        price_table: Price table for binned binned wind direction, wind
+            speed values (NDArrayFloat, optional).  Defaults to None.
+        compute_zero_freq_occurrence: Flag indicating whether to compute zero
+            frequency occurrences (bool, optional).  Defaults to False.
+
+    """
+
+    def __init__(
+        self,
+        wind_directions,
+        wind_speeds,
+        freq_table=None,
+        ti_table=None,
+        price_table=None,
+        compute_zero_freq_occurence=False,
+    ):
+        if not isinstance(wind_directions, np.ndarray):
+            raise TypeError("wind_directions must be a NumPy array")
+
+        if not isinstance(wind_speeds, np.ndarray):
+            raise TypeError("wind_directions must be a NumPy array")
+
+        # Save the wind speeds and directions
+        self.wind_directions = wind_directions
+        self.wind_speeds = wind_speeds
+
+        # If freq_table is not None, confirm it has correct dimension,
+        # otherwise initialze to uniform probability
+        if freq_table is not None:
+            if not freq_table.shape[0] == len(wind_directions):
+                raise ValueError("freq_table first dimension must equal len(wind_directions)")
+            if not freq_table.shape[1] == len(wind_speeds):
+                raise ValueError("freq_table second dimension must equal len(wind_speeds)")
+            self.freq_table = freq_table
+        else:
+            self.freq_table = np.ones((len(wind_directions), len(wind_speeds)))
+
+        # Normalize freq table
+        self.freq_table = self.freq_table / np.sum(self.freq_table)
+
+        # If TI table is not None, confirm dimension
+        # otherwise leave it None
+        if ti_table is not None:
+            if not ti_table.shape[0] == len(wind_directions):
+                raise ValueError("ti_table first dimension must equal len(wind_directions)")
+            if not ti_table.shape[1] == len(wind_speeds):
+                raise ValueError("ti_table second dimension must equal len(wind_speeds)")
+            self.ti_table = ti_table
+
+        else:
+            self.ti_table = None
```

Comment:
This is redundant

Reply:
Thanks, this and the one below are updated